internal/cloudsql/instance.go

```diff
@@ -166,32 +186,39 @@ func (i *Instance) scheduleRefresh(d time.Duration) *refreshResult {
 	res.ready = make(chan struct{})
 	res.timer = time.AfterFunc(d, func() {
 		ctx, cancel := context.WithTimeout(context.Background(), i.refreshTimeout)
-		res.md, res.tlsCfg, res.err = performRefresh(ctx, i.client, i.clientLimiter, i.connName, i.key)
+		res.md, res.tlsCfg, res.expiry, res.err = performRefresh(ctx, i.client, i.clientLimiter, i.connName, i.key)
 		cancel()
 
 		close(res.ready)
 		// Once the refresh is complete, update "current" with working result and schedule a new refresh
 		i.resultGuard.Lock()
 		defer i.resultGuard.Unlock()
-		// TODO: only replace cur result if it's not valid
-		i.cur = res
+		// if failed, scheduled the next refresh immediately
 		if res.err != nil {
-			// TODO: add a backoff on retries
-			// if failed, scheduled the next refresh immediately
 			i.next = i.scheduleRefresh(0)
+			// keep using current info unless it's no longer valid
+			// TODO: consider how to avoid supressing errors here
+			if !i.cur.IsValid() {
+				i.cur = res
```

Comment:
One option so that we don't have to lean on comments:

``` go
func WithErroredResult(err error) *refreshResult {
    return &refreshResult{error: err}
}
```

Reply:
Sounds good.